frame/election-provider-multi-phase/src/unsigned.rs

```diff
@@ -60,33 +73,134 @@ impl From<FeasibilityError> for MinerError {
 	}
 }
 
+/// Save a given call into OCW storage.
+fn save_solution<T: Config>(call: &Call<T>) -> Result<(), MinerError> {
+	let storage = StorageValueRef::persistent(&OFFCHAIN_CACHED_CALL);
+	match storage.mutate::<_, (), _>(|_| Ok(call.clone())) {
+		Ok(Ok(_)) => Ok(()),
+		Ok(Err(_)) => Err(MinerError::FailedToStoreSolution),
+		Err(_) => unreachable!("mutation function cannot fail; qed"),
+	}
+}
+
+/// Get a saved solution from OCW storage if it exists.
+fn restore_solution<T: Config>() -> Result<Call<T>, MinerError> {
+	StorageValueRef::persistent(&OFFCHAIN_CACHED_CALL)
+		.get()
+		.flatten()
+		.ok_or(MinerError::NoStoredSolution)
+}
+
+/// Clear a saved solution from OCW storage.
+pub(super) fn kill_solution<T: Config>() {
+	let mut storage = StorageValueRef::persistent(&OFFCHAIN_CACHED_CALL);
+	storage.clear();
+}
+
 impl<T: Config> Pallet<T> {
-	/// Mine a new solution, and submit it back to the chain as an unsigned transaction.
-	pub fn mine_check_and_submit() -> Result<(), MinerError> {
+	/// Attempt to restore a solution from cache. Otherwise, compute it fresh. Either way, submit
+	/// if our call's score is greater than that of the cached solution.
+	pub fn restore_or_compute_then_maybe_submit() -> Result<(), MinerError> {
+		log!(
+			info,
+			"OCW attempting to restore or compute an unsigned solution for the current election"
+		);
+
+		let (call, score) = restore_solution::<T>()
+			.and_then(|call| {
+				// ensure the cached call is still current before submitting
+				let current_round = Self::round();
+				match &call {
+					Call::submit_unsigned(solution, _) if solution.round == current_round => {
+						let score = solution.score.clone();
+
+						log!(
+							debug,
+							"restored a cached call for round {} with score {:?}",
+							current_round,
+							score,
+						);
+
+						Ok((call, score))
+					}
+					_ => Err(MinerError::SolutionOutOfDate),
+				}
+			})
+			.or_else::<MinerError, _>(|_| {
+				// if not present or cache invalidated, regenerate
+				let (call, score) = Self::mine_checked_call()?;
+				save_solution(&call)?;
+				Ok((call, score))
+			})?;
+
+		if Self::queued_solution().map_or(true, |q: ReadySolution<_>| is_score_better::<Perbill>(
```

Comment:
```suggestion
		// NOTE: for a solution that is just mined via `mine_checked_call`, we validate it via `unsigned_pre_dispatch_checks` and 
		// `feasibility_check`, both of which ensure that the score is already better. Thus, this
		// is kinda pointless for such cases. 
		if Self::queued_solution().map_or(true, |q: ReadySolution<_>| is_score_better::<Perbill>(
```

Or we just move this check to the `restore_solution::<T>().and_then(...)` body to avoid wasted work?

Reply:
Actually, I have a good reorg in mind, let me know what you think: 

Currently, some checks in `feasibility_check` is repeated in `unsigned_pre_dispatch_checks`. `unsigned_pre_dispatch_checks` itself is called both in `validate_unsigned` (which is used by the txpool), and in `pre_dispatch` (which is called exactly when the name says so: prior to _dispatch_, not _validation_). Currently, I am being superconservative: 

1. I repeat `unsigned_pre_dispatch_checks` is `submit_unsigned` itself which is redundant, because it is already called in `pre_dispatch`. 
2. Some individual components are checked twice: `DesiredTargets` is checked both in `feasibility_check` and `unsigned_pre_dispatch_checks`, and given the previous note is technically checked 3 times in a single dispatch pipeline: 

  1. in the call to `unsigned_pre_dispatch_checks` in `pre_dispatch`.
  2. in the call to `unsigned_pre_dispatch_checks` in `submit_unsigned`.
  3. lastly in the body of the `feasibility_check`.

We should rename `unsigned_pre_dispatch_checks` to a proper `pre_dispatch_checks` that checks:

1. phase 
2. round 
3. ~~desired_targets~~
4. score 

and everything else is done in `feasibility_check`.

The standard assumption is that all solutions **must** go through both of these eventually (and at least once), so we can deduplicate them: things that are checked in `pre_dispatch` need not be checked again in `feasibility_check`. Then the question is, what should be in `pre_dispatch` and what not?  

The role of `pre_dispatch` is this: feel the gap about **criteria that could change over time**. For example, if you are honest, you will always generate a solution that is passes `feasibility_check`. But, if you took an hour to compute it: 

1. The phase/round might have changed
2. `<QueuedSolution>` might now contain a better solution.

With this logic, the `pre_dispatch` should check for:

1. phase 
2. round (the combination of which guarantees that)
3. score 

to make sure if you were honest (used the correct snapshot), you will not submit a bad solution due to being out of date (round/phase) or the fact that a better solution was recently accepted (score). 

The reason that I historically stuffed `desired_targets` into `pre_dispatch` is: During weight/length trimming, we could remove a target, in which case the solution is now invalid, and we don't want to submit this. As a easy fix, I put it in `pre_dispatch_checks` because it was called in `validate_unsigned` and fixed this. But this is not a criterion that changes over time: once you generate a solution _and trim it_, if it has enough winner, it will stay so (until the round maybe changes). So calling into  `feasibility_check` in `mine_checked` is enough. 

So all in all: 

- The `mine_checked` should still call both of these just in case. 
- `feasibility_check` should still be called only in `submit_unsigned` (except for OCW code that calls it).  
- `pre_dispatch` should be called in both `validate` and `pre_dispatch` of `validate_unsigned`. 

Lastly, the cream on the case is that, here you can do something like: 

```
let maybe_call_and_score = restore_solution::<T>()
   .and_then(|call|  
       // `Ok` IFF call.solution is `pre_dispatch`. No need to check `feasibility` given all the notes above.
   ).or_else(|_| {
       // Same as before: mine_checked, save and return Ok
   })
```

This might sound too big, so maybe make it a follow-up, but the current layout is kinda confused as well. At the least, you can move the check to score only in `and_then`, and since now we are okay with duplicate checks, just call `unsigned_pre_dispatch_checks` in there to check round and score (with the other two as... bonus?)

Reply:
I like this idea in principle, but I think it's out of scope of this PR. I'd like to get this one closed out for now; we can work on reorganizing the checks later. I'll make a new issue out of this comment.

> At the least, you can move the check to score only in `and_then`

Tried that and it doesn't work; rust complains that it can't find `score` in that scope. My best guess is that `score` is only bound after the `restore_solution` -> `and_then` -> `or_else` chain completes.

Reply:
https://github.com/paritytech/substrate/issues/8641

Reply:
> Tried that and it doesn't work

I was dumb. Tried again and it works fine. 3f1109b67